--- a/kernel/runtime/ksud_integration.c
+++ b/kernel/runtime/ksud_integration.c
@@ -9,7 +9,6 @@
 #include <linux/fs.h>
 #include <linux/version.h>
 #include <linux/input-event-codes.h>
-#include <linux/kprobes.h>
 #include <linux/printk.h>
 #include <linux/types.h>
 #include <linux/uaccess.h>
@@ -23,8 +22,9 @@
 #include "runtime/ksud.h"
 #include "runtime/ksud_boot.h"
 #include "selinux/selinux.h"
-#include "hook/syscall_hook.h"
-#include "hook/syscall_event_bridge.h"
+
+DEFINE_STATIC_KEY_TRUE(ksu_is_init_rc_hook_enabled);
+DEFINE_STATIC_KEY_TRUE(ksu_is_input_hook_enabled);
 
 // clang-format off
 static const char KERNEL_SU_RC[] =
@@ -46,43 +46,25 @@
     "\n";
 // clang-format on
 
-static void stop_init_rc_hook();
-static void stop_execve_hook();
-
-static struct work_struct stop_input_hook_work;
-
-#define MAX_ARG_STRINGS 0x7FFFFFFF
-struct user_arg_ptr {
-#ifdef CONFIG_COMPAT
-    bool is_compat;
-#endif
-    union {
-        const char __user *const __user *native;
-#ifdef CONFIG_COMPAT
-        const compat_uptr_t __user *compat;
-#endif
-    } ptr;
-};
-
 static const char __user *get_user_arg_ptr(struct user_arg_ptr argv, int nr)
 {
-    const char __user *native;
+	const char __user *native;
 
 #ifdef CONFIG_COMPAT
-    if (unlikely(argv.is_compat)) {
-        compat_uptr_t compat;
+	if (unlikely(argv.is_compat)) {
+		compat_uptr_t compat;
 
-        if (get_user(compat, argv.ptr.compat + nr))
-            return ERR_PTR(-EFAULT);
+		if (get_user(compat, argv.ptr.compat + nr))
+			return ERR_PTR(-EFAULT);
 
-        return compat_ptr(compat);
-    }
+		return compat_ptr(compat);
+	}
 #endif
 
-    if (get_user(native, argv.ptr.native + nr))
-        return ERR_PTR(-EFAULT);
+	if (get_user(native, argv.ptr.native + nr))
+		return ERR_PTR(-EFAULT);
 
-    return native;
+	return native;
 }
 
 static int get_user_arg_ptr_len(struct user_arg_ptr argv)
@@ -148,80 +130,49 @@
 	return -EINVAL;
 }
 
-void ksu_execve_hook_ksud(const struct pt_regs *regs)
+int ksu_handle_execveat_ksud(int *fd, struct filename **filename_ptr,
+                             struct user_arg_ptr *argv,
+                             struct user_arg_ptr *envp, int *flags)
 {
-    struct user_arg_ptr argv, envp;
-    struct filename *filename;
-
-    filename = (struct filename *)PT_REGS_PARM2(regs);
-    argv.ptr.native = (const char __user *const __user *)PT_REGS_PARM3(regs);
-    envp.ptr.native = (const char __user *const __user *)PT_REGS_PARM4(regs);
-
-#ifdef CONFIG_COMPAT
-    argv.is_compat = is_compat_task();
-    envp.is_compat = is_compat_task();
-#endif
+    struct filename *filename = *filename_ptr;
 
     if (unlikely(strcmp(filename->name, "/system/bin/init") == 0)) {
-        if (get_user_arg_ptr_len(argv) == 1) {
-            on_post_fs_data();
-        }
+        if (get_user_arg_ptr_len(*argv) == 1) {
+			on_post_fs_data();
+		}
     }
+    return 0;
 }
 
-extern struct kprobe input_event_kp;
-
-static void do_stop_input_hook(struct work_struct *work)
-{
-    unregister_kprobe(&input_event_kp);
-}
-
-static void stop_init_rc_hook()
-{
-    ksu_syscall_table_unhook(__NR_read);
-    ksu_syscall_table_unhook(__NR_fstat);
-    pr_info("unregister init_rc syscall hook\n");
-}
-
-void ksu_stop_input_hook_runtime(void)
-{
-    static bool input_hook_stopped = false;
-    if (input_hook_stopped) {
-        return;
-    }
-    input_hook_stopped = true;
-    bool ret = schedule_work(&stop_input_hook_work);
-    pr_info("unregister input kprobe: %d!\n", ret);
-}
-
-static int vfs_read_handler_pre(struct kprobe *p, struct pt_regs *regs)
+void ksu_handle_vfs_read_init_rc(struct file *file, char __user *buf, size_t count, loff_t *pos)
 {
-    struct file *file = (struct file *)PT_REGS_PARM1(regs);
-    char __user *buf = (char __user *)PT_REGS_PARM2(regs);
-    size_t count = (size_t)PT_REGS_PARM3(regs);
-    loff_t *pos = (loff_t *)PT_REGS_PARM4(regs);
+	if (!static_branch_likely(&ksu_is_init_rc_hook_enabled))
+		return;
 
-    if (is_init_rc(file)) {
-        if (*pos == 0) {
-            // we only hook the first read
-            if (count >= sizeof(KERNEL_SU_RC)) {
-                if (copy_to_user(buf, KERNEL_SU_RC, sizeof(KERNEL_SU_RC))) {
-                    pr_err("copy_to_user failed\n");
-                }
-                *pos += sizeof(KERNEL_SU_RC);
-            }
-        }
-    }
-
-    return 0;
+	if (is_init_rc(file)) {
+		if (*pos == 0) {
+			// we only hook the first read
+			if (count >= sizeof(KERNEL_SU_RC)) {
+				if (copy_to_user(buf, KERNEL_SU_RC, sizeof(KERNEL_SU_RC))) {
+					pr_err("copy_to_user failed\n");
+				}
+				*pos += sizeof(KERNEL_SU_RC);
+			}
+		}
+	}
 }
 
-static int ksu_sys_read(const struct pt_regs *regs)
+void ksu_handle_vfs_stat_init_rc(struct file *file, u32 *kstat_size_ptr)
 {
-    struct file *file = fget(PT_REGS_PARM1(regs));
-    char __user *buf = (char __user *)PT_REGS_PARM2(regs);
-    size_t count = (size_t)PT_REGS_PARM3(regs);
-    loff_t *pos = (loff_t *)PT_REGS_PARM4(regs);
+	static const int new_size = sizeof(KERNEL_SU_RC) - 1;
+
+	if (!static_branch_likely(&ksu_is_init_rc_hook_enabled))
+		return;
+
+	if (!file)
+		return;
+
+	if (is_init_rc(file)) {
+		pr_info("stat init.rc");
+		pr_info("adding ksu_rc_len: %lld -> %lld", *kstat_size_ptr, new_size);
+		*kstat_size_ptr = new_size;
+	}
+	fput(file);
+}
 
-    if (!file) {
-        return -EBADF;
-    }
+void ksu_handle_vfs_read_input(struct file *file, char __user *buf, size_t count, loff_t *pos)
+{
+	if (!static_branch_likely(&ksu_is_input_hook_enabled))
+		return;
 
-    if (is_init_rc(file)) {
-        if (*pos == 0) {
-            // we only hook the first read
-            if (count >= sizeof(KERNEL_SU_RC)) {
-                if (copy_to_user(buf, KERNEL_SU_RC, sizeof(KERNEL_SU_RC))) {
-                    pr_err("copy_to_user failed\n");
-                }
-                *pos += sizeof(KERNEL_SU_RC);
-                fput(file);
-                return sizeof(KERNEL_SU_RC);
-            }
-        }
-    }
-    fput(file);
-    return orig_sys_read(regs);
-}
-
-static int ksu_sys_fstat(const struct pt_regs *regs)
-{
-    struct file *file = fget(PT_REGS_PARM1(regs));
-    struct kstat __user *statbuf = (struct kstat __user *)PT_REGS_PARM2(regs);
-    static const int new_size = sizeof(KERNEL_SU_RC) - 1;
-    int ret;
-
-    if (!file) {
-        return -EBADF;
-    }
-
-    ret = orig_sys_fstat(regs);
-
-    if (is_init_rc(file)) {
-        pr_info("stat init.rc");
-        if (copy_to_user(&statbuf->size, &new_size, sizeof(new_size))) {
-            pr_err("copy_to_user failed\n");
-        }
-    }
-    fput(file);
-    return ret;
+	if (is_input_device(file)) {
+		// dummy
+	}
 }
 
 // ksud: module support
 void __init ksu_ksud_init()
 {
-    int ret;
-
-    ksu_syscall_table_hook(__NR_read, ksu_sys_read, &orig_sys_read);
-    ksu_syscall_table_hook(__NR_fstat, ksu_sys_fstat, &orig_sys_fstat);
-
-    ret = register_kprobe(&input_event_kp);
-    pr_info("ksud: input_event_kp: %d\n", ret);
-
-    INIT_WORK(&stop_input_hook_work, do_stop_input_hook);
 }
 
 void __exit ksu_ksud_exit()
 {
-    // TODO:
-    // this should be done before unregister vfs_read_kp
-    // stop_init_rc_hook();
-    unregister_kprobe(&input_event_kp);
 }
