diff --git a/kernel/hook/setuid_hook.h b/kernel/hook/setuid_hook.h
index b8486238..75dea4e9 100644
--- a/kernel/hook/setuid_hook.h
+++ b/kernel/hook/setuid_hook.h
@@ -5,6 +5,9 @@
 #include <linux/types.h>
 
 void ksu_setuid_hook_init(void);
-void ksu_setuid_hook_exit(void);
+void __exit ksu_setuid_hook_exit(void);
+
+// Handler functions for hook_manager
+int ksu_handle_setresuid(uid_t ruid, uid_t euid, uid_t suid);
 
 #endif
