diff --git a/kernel/supercall/supercall.h b/kernel/supercall/supercall.h
index fbedd11f..483e5069 100644
--- a/kernel/supercall/supercall.h
+++ b/kernel/supercall/supercall.h
@@ -18,6 +18,7 @@ struct ksu_ioctl_cmd_map {
 
 // Install KSU fd to current process
 int ksu_install_fd(void);
+int ksu_supercall_reboot_handler(void __user **arg);
 
 void ksu_supercalls_init(void);
 void ksu_supercalls_exit(void);
