--- a/kernel/runtime/ksud.h
+++ b/kernel/runtime/ksud.h
@@ -8,7 +10,20 @@
 void ksu_ksud_init();
 void ksu_ksud_exit();
 
-void ksu_execve_hook_ksud(const struct pt_regs *regs);
-void ksu_stop_input_hook_runtime(void);
+#define MAX_ARG_STRINGS 0x7FFFFFFF
+struct user_arg_ptr {
+#ifdef CONFIG_COMPAT
+    bool is_compat;
+#endif
+    union {
+        const char __user *const __user *native;
+#ifdef CONFIG_COMPAT
+        const compat_uptr_t __user *compat;
+#endif
+    } ptr;
+};
 
+int ksu_handle_execveat_ksud(int *fd, struct filename **filename_ptr,
+                             struct user_arg_ptr *argv,
+                             struct user_arg_ptr *envp, int *flags);
 #endif
