diff --git a/kernel/policy/app_profile.c b/kernel/policy/app_profile.c
index c3db708c..b8d1ae3d 100644
--- a/kernel/policy/app_profile.c
+++ b/kernel/policy/app_profile.c
@@ -169,10 +169,10 @@ int escape_with_root_profile(void)
     // setup capabilities
     // we need CAP_DAC_READ_SEARCH becuase `/data/adb/ksud` is not accessible for non root process
     // we add it here but don't add it to cap_inhertiable, it would be dropped automaticly after exec!
-    u64 cap_for_ksud = profile->capabilities.effective | CAP_DAC_READ_SEARCH;
+    u64 cap_for_ksud = profile->capabilities.effective | (1ULL << CAP_DAC_READ_SEARCH);
     memcpy(&cred->cap_effective, &cap_for_ksud, sizeof(cred->cap_effective));
-    memcpy(&cred->cap_permitted, &profile->capabilities.effective, sizeof(cred->cap_permitted));
-    memcpy(&cred->cap_bset, &profile->capabilities.effective, sizeof(cred->cap_bset));
+    memcpy(&cred->cap_permitted, &cap_for_ksud, sizeof(cred->cap_permitted));
+    memcpy(&cred->cap_bset, &cap_for_ksud, sizeof(cred->cap_bset));
 
     setup_groups(profile, cred);
     setup_selinux(profile->selinux_domain, cred);
@@ -181,10 +181,6 @@ int escape_with_root_profile(void)
 
     disable_seccomp();
 
-    for_each_thread (p, t) {
-        ksu_set_task_tracepoint_flag(t);
-    }
-
     setup_mount_ns(profile->namespaces);
     ksu_put_root_profile(profile);
     return 0;
