diff --git a/kernel/hook/setuid_hook.c b/kernel/hook/setuid_hook.c
index d3aef3c1..3515f944 100644
--- a/kernel/hook/setuid_hook.c
+++ b/kernel/hook/setuid_hook.c
@@ -93,10 +93,21 @@ int ksu_handle_setresuid(uid_t ruid, uid_t euid, uid_t suid)
         ksu_clear_task_tracepoint_flag_if_needed(current);
     }
 
-    // Handle kernel umount
-    ksu_handle_umount(old_uid, new_uid);
+	if (unlikely(ksu_su_compat_enabled)) {
+		ksu_handle_setresuid(old_cred->uid.val, new_cred->uid.val, 0);
+	}
 
-    return 0;
+	if (unlikely(new_cred->uid.val == 0)) {
+		susfs_handle_setresuid(old_cred->uid.val, new_cred->uid.val, 0);
+		goto do_umount;
+	}
+
+	return 0;
+
+do_umount:
+	ksu_handle_umount(old_cred->uid.val, new_cred->uid.val);
+
+	return 0;
 }
 
 void __init ksu_setuid_hook_init(void)
