diff --git a/kernel/Kbuild b/kernel/Kbuild
index eb87e088..ac821722 100644
--- a/kernel/Kbuild
+++ b/kernel/Kbuild
@@ -1,4 +1,15 @@
+## For susfs stuff ##
+ifeq ($(shell test -e $(srctree)/fs/susfs.c; echo $$?),0)
+$(eval SUSFS_VERSION=$(shell cat $(srctree)/include/linux/susfs.h | grep -E '^#define SUSFS_VERSION' | cut -d' ' -f3 | sed 's/"//g'))
+$(info )
+$(info -- SUSFS_VERSION: $(SUSFS_VERSION))
+ccflags-y += -DCONFIG_KSU_SUSFS
+else
+$(info -- You have not integrated susfs in your kernel yet.)
+$(info -- Read: https://gitlab.com/simonpunk/susfs4ksu)
+endif
+
 kernelsu-objs := core/init.o
 kernelsu-objs += feature/kernel_umount.o
 kernelsu-objs += feature/sucompat.o
