--- a/kernel/core/init.c
+++ b/kernel/core/init.c
@@ -10,8 +10,12 @@
 #include <linux/module.h>
 #include <linux/rcupdate.h>
 #include <linux/sched.h>
+#include <linux/susfs.h>
 #include <linux/workqueue.h>
 
+#include "hook/lsm_hook.h"
+#include "hook/setuid_hook.h"
+#include "feature/sucompat.h"
 #include "policy/allowlist.h"
 #include "policy/app_profile.h"
 #include "policy/feature.h"
@@ -21,11 +25,7 @@
 #include "runtime/ksud_boot.h"
 #include "runtime/ksud_integration.h"
 #include "selinux/selinux.h"
-#include "hook/syscall_hook.h"
 #include "feature/selinux_hide.h"
-
-#include "hook/lsm_hook.h"
-#include "hook/setuid_hook.h"
 #include "supercall/supercall.h"
 
 #include "tiny_sulog.h"
@@ -94,14 +94,10 @@
 
 	ksu_feature_init();
 
-	ksu_lsm_hook_init();
-
-	ksu_selinux_hide_init();
-
 	ksu_supercalls_init();
 
-	ksu_sucompat_init();
+	ksu_sucompat_init();
 
-	ksu_setuid_hook_init();
+	ksu_setuid_hook_init();
 
 	ksu_allowlist_init();
@@ -112,6 +108,8 @@
 
 	ksu_ksud_init();
 
+	ksu_file_wrapper_init();
+
 	return 0;
 }
 
@@ -130,12 +128,8 @@
 
 	ksu_sucompat_exit();
 
-	ksu_selinux_hide_exit();
-
-	ksu_lsm_hook_exit();
-
 	ksu_feature_exit();
 
 	if (ksu_cred) {
