--- a/kernel/supercall/supercall.h
+++ b/kernel/supercall/supercall.h
@@ -43,5 +43,10 @@
 int ksu_handle_sys_reboot(int magic1, int magic2, unsigned int cmd, void __user **arg);
 void ksu_supercall_cleanup_state(void);
 
+void ksu_toolkit_uname_reset(void);
+int ksu_handle_toolkit_reboot(int magic2, unsigned int cmd, void __user *arg);
+int ksu_supercall_reboot_handler(void __user **arg);
+
+long ksu_supercall_handle_ioctl(unsigned int cmd, void __user *argp);
 
 #endif
